### PR TITLE
wip

### PR DESCRIPTION
**Pull request by musgravejw**
_wip_ for https://github.com/musgravejw/test-html-parser/issues/1

These are some of the easiest parts of the HTML grammar to parse, because they don't have any attributes.

The parser in it's current state does not address the non-context free grammar.  

With this approach, every production must be checked recursively at multiple points in the grammar.  Haven't figured out a good way to do this yet.

### CODE REVIEW

File: scanner.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#include <string.h>
+#include <map>
+
+class Scanner {
+  private:
+    std::map<string, string> Grammar;
+
+  public:
+    Scanner() {
```
Comment (domfarolino) [inline scanner.cpp]: We might want to initialize the Grammar map in the initialization list (there should be `{{...}}` syntax to do this). That might be a little cleaner and more performant.

Comment (domfarolino) [reply scanner.cpp]: Also I might call it "Terminals" instead of "Grammar" but that's a nit.

Comment (musgravejw) [reply scanner.cpp]: @domfarolino Yeah, my C++ is probably terrible.  Also new to C++ 14/17.

File: scanner.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#include <string.h>
+#include <map>
+
+class Scanner {
+  private:
+    std::map<string, string> Grammar;
```
Comment (domfarolino) [inline scanner.cpp]: I'd make this an `unordered_map` too.

File: scanner.cpp (hunk 1 above)
Comment (domfarolino) [reply scanner.cpp]: This article (http://thispointer.com/different-ways-to-initialize-an-unordered_map/) might be a good reference for the initialization. I don't think it is C++ 14/17 specific (at least hopefully!)

**Review by domfarolino — CHANGES_REQUESTED** [review]

**Pull request closed without merging**